linkerd/http/retry/src/lib.rs

```diff
@@ -81,31 +83,45 @@ struct Metrics {
 
 // === impl NewHttpRetry ===
 
-impl<P, L: Clone, N> NewHttpRetry<P, L, N> {
-    pub fn layer(metrics: MetricFamilies<L>) -> impl layer::Layer<N, Service = Self> + Clone {
+impl<P, X, L: Clone, F: Clone, N> NewHttpRetry<P, L, F, X, N> {
+    pub fn layer_via_mk(
+        mk_extract: F,
+        metrics: MetricFamilies<L>,
+    ) -> impl tower::layer::Layer<N, Service = Self> + Clone {
```

Comment:
:point_up: bike-shedding welcome. other `layer_via` functions conventionally receive an `X` extractor, in our case we want a function that takes a reference to the `NewService<T>` target (_a matched route_) to construct an `X` that can later extract our `L` labels (_including a hostname_) provided the `Service<T>` target (_a request_).

in a sense this is almost like a `NewExtractParam<..>`, if we can compare it to `NewService`. see the commit message for a little bit more information.

Reply:
It took me a bit to fully grok this (the several layers of tower abstraction are always a bit of a pain to punch through), but I think this is probably the best option given the naming scheme we already have to work with.

Reply:
https://github.com/linkerd/linkerd2-proxy/pull/3258#discussion_r1794142955 for posterity, some of the changes in this thread helped clarify the layers of :cake: :sparkles: abstraction extraction